src/main/scala/com/campudus/tableaux/router/auth/permission.scala

```diff
@@ -0,0 +1,305 @@
+    }
+  }
+}
+
+case object Create extends Action { override val name = "create" }
+case object View extends Action { override val name = "view" }
+case object Edit extends Action { override val name = "edit" } // only for media
+case object Delete extends Action { override val name = "delete" }
+case object EditDisplayProperty extends Action { override val name = "editDisplayProperty" }
+case object CreateRow extends Action { override val name = "createRow" }
+case object DeleteRow extends Action { override val name = "deleteRow" }
+case object EditStructureProperty extends Action { override val name = "editStructureProperty" }
+case object EditCellAnnotation extends Action { override val name = "editCellAnnotation" }
+case object EditRowAnnotation extends Action { override val name = "editRowAnnotation" }
+case object ViewCellValue extends Action { override val name = "viewCellValue" }
+case object EditCellValue extends Action { override val name = "editCellValue" }
+
+sealed trait Scope {
+  val name: String
+  override def toString: String = name
+}
+
+object Scope {
+
+  def apply(scope: String): Scope = {
+    scope match {
+      case ScopeTable.name => ScopeTable
+      case ScopeColumn.name => ScopeColumn
+      case ScopeRow.name => ScopeRow
+      case ScopeCell.name => ScopeCell
+      case ScopeMedia.name => ScopeMedia
+      case _ => throw new IllegalArgumentException(s"Invalid argument for PermissionType $scope")
+    }
+  }
+}
+
+case object ScopeTable extends Scope { override val name = "table" }
+case object ScopeColumn extends Scope { override val name = "column" }
+case object ScopeRow extends Scope { override val name = "row" }
+case object ScopeCell extends Scope { override val name = "cell" }
+case object ScopeMedia extends Scope { override val name = "media" }
+
+object ConditionContainer {
+
+  def apply(jsonObjectOrNull: JsonObject): ConditionContainer = {
+
+    val jsonObject: JsonObject = Option(jsonObjectOrNull).getOrElse(Json.emptyObj())
+
+    val ct: ConditionOption = Option(jsonObject.getJsonObject("table")).map(ConditionTable).getOrElse(NoneCondition)
+    val cc: ConditionOption = Option(jsonObject.getJsonObject("column")).map(ConditionColumn).getOrElse(NoneCondition)
+    val cl: ConditionOption =
+      Option(Json.obj("langtag" -> jsonObject.getString("langtag"))).map(ConditionLangtag).getOrElse(NoneCondition)
+
+    new ConditionContainer(ct, cc, cl)
+  }
+}
+
+case class ConditionContainer(conditionTable: ConditionOption,
+                              conditionColumn: ConditionOption,
+                              conditionLangtag: ConditionOption) {}
+
+abstract class ConditionOption(jsonObject: JsonObject) {
+  val conditionMap: Map[String, String] = toMap(jsonObject)
+
+  protected def toMap(jsonObject: JsonObject) = {
+    jsonObject.asMap.toMap.asInstanceOf[Map[String, String]]
+  }
+
+  // TODO remove ugly method overloading, but how?
```

Comment:
isMatching accepts parameter of any type and inside the method you match for the type and act regardingly.
Not the best solution either...

Reply:
solved with a container